File: GameEngine_IOS/GameEngine/Shaders/SkyBox/SkyBoxShaderManager.m


#import "SkyBoxShaderManager.h"

/**
 * Manager of the shader files that are going to be load to render the 3D
 */
@implementation SkyBoxShaderManager

{
@private
    
    // Uniform index.
    enum
    {
        PROJECTION_MATRIX,
        VIEW_MATRIX,
        NUM_UNIFORMS
    };
    GLint uniforms[NUM_UNIFORMS];
    
    // Attribute index.
    enum
    {
        ATTRIB_VERTEX,
        ATTRIB_NORMAL,
        NUM_ATTRIBUTES
    };
}

/**
 * Initializor of the game shader where the vertex and fragment shader of
 * the game engine are loaded
 *
 */
- (id)init {
    self = [super init: @"sky_box_vertex_shader" : @"sky_box_fragment_shader"];
    return self;
}

/**
 * Called to bind the attributes to the program shader
 */
- (void) bindAttributes {
    [super bindAttribute : LOCATION_ATTR_ID: "position"];
}

/**
 * Called to ensure that all the shader managers get their uniform locations
 */
- (void) getAllUniformLocations {
    @autoreleasepool {
        NSDictionary * uniformsDic = [NSDictionary dictionaryWithObjectsAndKeys:
                                      @"projectionMatrix", [NSNumber numberWithInt:PROJECTION_MATRIX],
                                      @"viewMatrix", [NSNumber numberWithInt: VIEW_MATRIX],
                                      nil];
        for(id key in uniformsDic) {
            int iKey = [key intValue];
            const char *shaderVariable = [(uniformsDic[key]) UTF8String];
            uniforms[iKey] = [super getUniformLocation: shaderVariable];
            if(uniforms[iKey] < 0) {
                NSLog(@"Problems getting %@'s location in the shader", uniformsDic[key]);
            }
        }
    }
}


/**
 * Load the projection matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadProjectionMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[PROJECTION_MATRIX]: matrix];
}

/**
 * Load the view matrix
 *
 * @param matrix
 *            the matrix to be loaded
 */
- (void) loadViewMatrix : (GLTransformation*) matrix {
    [super loadMatrix:  uniforms[VIEW_MATRIX]: matrix];
}

@end

File: GameEngine_IOS/GameEngine/Shaders/SkyBox/SkyBoxShaderManager.swift
import Foundation

open class SkyBoxShaderManager : ShaderManager {
    
    
    
    /**
    * Initializor of the game shader where the vertex and fragment shader of
    * the game engine are loaded
    *
    */
    public init() {
        super.init(vertexFile: "sky_box_vertex_shader" , fragmentFile: "sky_box_fragment_shader", numberOfUniforms: TSkyBoxUniform.numOfSkyBoxLocations.rawValue)
    }
    
    /*
    * Called to provide the attributes that are goint to get bound to the shader
    */
    override internal func getAttributes() -> Dictionary<Int, String>! {
        let attributesDic : Dictionary<Int, String> = [
            TSkyBoxAttribute.position.rawValue : "\(TSkyBoxAttribute.position)"
        ];
        return attributesDic;
    }
    
    /**
    * Called to provide the uniform locations that are going to get bound to the shader
    */
    override internal func getUniformLocations() -> Dictionary<NSInteger, String>! {
        let uniformsDic : Dictionary<NSInteger, String> = [
            TSkyBoxUniform.projectionMatrix.rawValue : "\(TSkyBoxUniform.projectionMatrix)",
            TSkyBoxUniform.viewMatrix.rawValue : "\(TSkyBoxUniform.viewMatrix)"
        ]
        
        return uniformsDic;
    }
    
    /**
    * Load the projection matrix
    *
    * @param matrix
    *            the matrix to be loaded
    */
    open func loadProjectionMatrix (_ matrix : GLTransformation) {
        super.loadMatrix(uniforms[TSkyBoxUniform.projectionMatrix.rawValue],matrix: matrix);
    }
    
    /**
    * Load the view matrix
    *
    * @param matrix
    *            the matrix to be loaded
    */
    open func loadViewMatrix (_ matrix : GLTransformation) {
        matrix.setTranslation(x: 0.0, y: 0.0, z: 0.0)
        super.loadMatrix(uniforms[TSkyBoxUniform.viewMatrix.rawValue],matrix: matrix);
    }
}
